printrun/printcore.py

```diff
@@ -350,24 +350,28 @@ def _readline(self):
             self.logError(_("Got rubbish reply from %s at baudrate %s:") % (self.port, self.baud) +
                               "\n" + _("Maybe a bad baudrate?"))
             return None
-        except SelectError as e:
-            if 'Bad file descriptor' in e.args[1]:
-                self.logError(_("Can't read from printer (disconnected?) (SelectError {0}): {1}").format(e.errno, decode_utf8(e.strerror)))
-                return None
-            else:
-                self.logError(_("SelectError ({0}): {1}").format(e.errno, decode_utf8(e.strerror)))
-                raise
         except SerialException as e:
             self.logError(_("Can't read from printer (disconnected?) (SerialException): {0}").format(decode_utf8(str(e))))
             return None
         except socket.error as e:
             self.logError(_("Can't read from printer (disconnected?) (Socket error {0}): {1}").format(e.errno, decode_utf8(e.strerror)))
             return None
-        except OSError as e:
-            if e.errno == errno.EAGAIN:  # Not a real error, no data was available
-                return ""
-            self.logError(_("Can't read from printer (disconnected?) (OS Error {0}): {1}").format(e.errno, e.strerror))
-            return None
+        except (OSError, SelectError) as e:
+            # OSError and SelectError are the same thing since python 3.3
+            if self.printer_tcp:
```

Comment:
Why not check if e is a SelectError rather than assuming here?

Reply:
SelectError must go to if branch, OSError must go to else branch.
When OSError is SelectError, then OSError will go to if branch, and we do not want this.
Type testing will not work, so I figure the semantics by other means.
I believe this will work both before and after python3.3. I did not made the effort to check whether we claim to support python3.3, but I am also not sure if this type aliasing is actually made in python3.3. Different python distributions/implementations further add ambiguity.
My strategy is to reuse the existing handlers and only dispatch between them because I did not reproduce the handled exceptions and did not verify their actual signatures.

Reply:
I am thinking about replacing if self.printer_tcp with if type(e) is SelectError. Shouldn't that work?

Reply:
Either way, feel free to merge as is. If it works like this I'm fine with it.

Reply:
select.py
```
from sys import OSError as error
```

```
from select import error as SelectError
```

I you replace it, OSError from serial printers will be handled in the if branch, which never returns "" to signify "no data, try again later".
So when the printer is idle, it will be interpreted as connection failure.